lib/vrf.c

```diff
@@ -424,32 +446,86 @@ int vrf_socket(int domain, int type, int protocol, vrf_id_t vrf_id)
 	return ret;
 }
 
-/* vrf CLI commands */
-DEFUN_NOSH (vrf,
-       vrf_cmd,
-       "vrf NAME",
-       "Select a VRF to configure\n"
-       "VRF's name\n")
+extern int vrf_is_backend_ns(void)
+{
+	return vrf_backend_netns;
+}
+
+void vrf_configure_backend(int ns_on)
+{
+	vrf_backend_netns = ns_on;
+}
+
+int vrf_handler_create(struct vty *vty, const char *vrfname, struct vrf **vrf)
 {
-	int idx_name = 1;
-	const char *vrfname = argv[idx_name]->arg;
 	struct vrf *vrfp;
 
 	if (strlen(vrfname) > VRF_NAMSIZ) {
-		vty_out(vty,
-			"%% VRF name %s is invalid: length exceeds "
-			"%d characters\n",
-			vrfname, VRF_NAMSIZ);
+		if (vty)
+			vty_out(vty,
+			    "%% VRF name %s is invalid: length exceeds "
+			    "%d characters\n", vrfname, VRF_NAMSIZ);
+		else
+			zlog_warn("%% VRF name %s is invalid: length exceeds "
+				  "%d characters\n", vrfname, VRF_NAMSIZ);
 		return CMD_WARNING_CONFIG_FAILED;
 	}
 
 	vrfp = vrf_get(VRF_UNKNOWN, vrfname);
 
-	VTY_PUSH_CONTEXT(VRF_NODE, vrfp);
+	if (vty)
+		VTY_PUSH_CONTEXT(VRF_NODE, vrfp);
 
+	if (vrf)
+		*vrf = vrfp;
 	return CMD_SUCCESS;
 }
 
+int vrf_is_mapped_on_netns(struct vrf *vrf)
+{
+	if (!vrf || vrf->data.ns_name[0] == '\0')
+		return 0;
+	if (vrf->vrf_id == NS_DEFAULT)
+		return 0;
+	return 1;
+}
+
+int vrf_switch_to_netns(struct vrf *vrf)
+{
+	char *name;
+
+	if (!vrf || vrf->data.ns_name[0] == '\0')
+		return -1;
+	if (vrf->vrf_id == NS_DEFAULT)
+		return -1;
+	name = ns_netns_pathname(NULL, vrf->data.ns_name);
+	if (debug_vrf)
+		zlog_debug("VRF_SWITCH: %s(%u)", name, vrf->vrf_id);
+	return ns_switch_to_netns (name);
+}
+
+int vrf_switchback_to_initial (void)
+{
+	int ret = ns_switchback_to_initial ();
+
+	if (ret == 0 && debug_vrf)
+		zlog_debug("VRF_SWITCHBACK");
+	return ret;
```

Comment:
it can occur, because the place where vrf_switchback_to_initial is called may be wrong.
I mean, there is no reason that it aborts if the caller ensured that he called switch/switchback properly.

Reply:
My concern is the following: if we fail to switch back to the initial netns, then all subsequent operations will be performed on a different netns. It's a recipe for disaster. Are we going to just ignore this? IMO it's better to fail earlier than keep running in an inconsistent state. We have to keep in mind that this is mostly a theoretical problem since `ns_switchback_to_initial()` shouldn't fail in normal conditions.